frontend-utils/src/player_options/read.rs

```diff
@@ -0,0 +1,751 @@
+        assert_eq!(
+            vec![ParseWarning::UnsupportedValue {
+                value: "invalid".to_string(),
+                path: "base_url".to_string(),
+            }],
+            result.warnings
+        );
+
+        let result = read("base_url = \"file:///example/path/\"");
+        assert_eq!(
+            &PlayerOptions {
+                base: Some(Url::parse("file:///example/path/").unwrap()),
+                ..Default::default()
+            },
+            result.values()
+        );
+        assert_eq!(Vec::<ParseWarning>::new(), result.warnings);
+    }
+
+    #[test]
+    fn quality() {
+        let result = read("quality = false");
+        assert_eq!(&PlayerOptions::default(), result.values());
+        assert_eq!(
+            vec![ParseWarning::UnexpectedType {
+                expected: "string",
+                actual: "boolean",
+                path: "quality".to_string()
+            }],
+            result.warnings
+        );
+
+        let result = read("quality = \"fabulous\"");
+        assert_eq!(&PlayerOptions::default(), result.values());
+        assert_eq!(
+            vec![ParseWarning::UnsupportedValue {
+                value: "fabulous".to_string(),
+                path: "quality".to_string(),
+            }],
+            result.warnings
+        );
+
+        fn assert_variant(variant: &str, quality: StageQuality) {
+            let result = read(&format!("quality = \"{variant}\""));
+            assert_eq!(
+                &PlayerOptions {
+                    quality: Some(quality),
+                    ..Default::default()
+                },
+                result.values()
+            );
+            assert_eq!(Vec::<ParseWarning>::new(), result.warnings);
+
+            // Check case-sensitivity.
+            let variant = variant.to_ascii_uppercase();
+            let result = read(&format!("quality = \"{variant}\""));
+            assert_eq!(
+                &PlayerOptions {
+                    quality: None,
+                    ..Default::default()
+                },
+                result.values()
+            );
+            assert_eq!(
+                vec![ParseWarning::UnsupportedValue {
+                    value: variant,
+                    path: "quality".to_string()
+                }],
+                result.warnings
+            );
+        }
+        assert_variant("high", StageQuality::High);
+        assert_variant("medium", StageQuality::Medium);
+        assert_variant("low", StageQuality::Low);
+
+        // TODO: Remove or add to all tests with all variants? (seems like a lot of code).
+        // Values are case-sensitive.
+        let result = read("quality = \"High\"");
+        assert_eq!(&PlayerOptions::default(), result.values());
+        assert_eq!(
+            vec![ParseWarning::UnsupportedValue {
+                value: "High".to_string(),
+                path: "quality".to_string(),
+            }],
+            result.warnings
+        );
```

Comment:
I forgot to remove this when I merged case-sensitivity check into assert_variant. Will remove later.